### PR TITLE
stm32wl5: add flash progmem driver support

### PR DESCRIPTION
**Pull request by mlyszczek**
## Summary
These patches add support for progmem driver to use internal FLASH memory to hold user's data. Support to partition FLASH to up to 4 partitions has beed added to nucleo-wl55jc board. User can also configure which filesystem to use for each partition.

## Impact
None, new code

## Testing
Compilation + checking if all 4 partitions initialize properly and automount at board startup. Checked if data is retained between reboots and reflashing.

### CODE REVIEW

File: arch/arm/src/stm32wl5/stm32wl5_flash.c [hunk 1]
```diff
@@ -359,6 +343,31 @@ ssize_t up_progmem_eraseblock(size_t block)
     }
 }
 
+ssize_t up_progmem_ispageerased(size_t page)
```
Comment (xiaoxiang781216) [inline arch/arm/src/stm32wl5/stm32wl5_flash.c:346]: why change the order of up_progmem_ispageerased and up_progmem_eraseblock? it make the change is hard to compare.

File: boards/arm/stm32wl5/nucleo-wl55jc/src/stm32_flash.c [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+{
+    int         size; /* partition size in pages */
+    const char *name; /* name of the partition */
+    const char *mnt;  /* mount point for the partition */
+    const char *fs;   /* fs type (smart, raw, nxffs) */
+};
+
+/* partition table, first entry *must always* be program flash memory */
+
+static struct part_table part_table[] =
```
Comment (xiaoxiang781216) [inline boards/arm/stm32wl5/nucleo-wl55jc/src/stm32_flash.c:76]: add const

File: arch/arm/src/stm32wl5/stm32wl5_flash.c (hunk 1 above)
Comment (mlyszczek) [reply arch/arm/src/stm32wl5/stm32wl5_flash.c:346]: Orginally stm32wl5_flash.c was based on code copied from stm32l5, but it turned out stm32l4 is more compatible with stm32wl5, so I copied over file from l4 and then adapted the code to work with wl5. I think we should consider files in arch/* as new files, not diffs.

Comment (xiaoxiang781216) [reply arch/arm/src/stm32wl5/stm32wl5_flash.c:346]: OK

**Comment by xiaoxiang781216** [comment]
@onegray do you want we wait you finish review before merging?

**Review by onegray — CHANGES_REQUESTED** [review]

File: arch/arm/src/stm32wl5/hardware/stm32wl5_flash.h [hunk 3]
```diff
@@ -131,174 +129,174 @@
...
-#define FLASH_CR_FSTPG              (1 << 23)               /* Bit 23: Fast programming */
-#define FLASH_CR_EOPIE              (1 << 24)               /* Bit 24: End of operation interrupt enable */
-#define FLASH_CR_ERRIE              (1 << 25)               /* Bit 25: Error interrupt enable */
-#define FLASH_CR_RDERRIE            (1 << 26)               /* Bit 26: PCROP read error interrupt enable */
-#define FLASH_CR_OBL_LAUNCH         (1 << 27)               /* Bit 27: Option Byte Loading */
-#define FLASH_CR_OPTLOCK            (1 << 30)               /* Bit 30: Option Lock */
-#define FLASH_CR_LOCK               (1 << 31)               /* Bit 31: Lock */
+#define FLASH_CR_START             (1 << 16) /* Bit 16: Start Erase */
+#define FLASH_CR_OPTSTRT           (1 << 17) /* Bit 17: Options modification Start */
+#define FLASH_CR_FSTPG             (1 << 23) /* Bit 23: Fast programming */
```
[the remark this answers is not in the document]
Comment (onegray) [reply arch/arm/src/stm32wl5/hardware/stm32wl5_flash.h:208]: @xiaoxiang781216 I've finished and have proposed minor changes. I just forget to submit my review, thanks.

**Comment by onegray** [comment]
> do you want we wait you finish review before merging?

@xiaoxiang781216 I've finished and have proposed minor changes. I just forget to submit my review, thanks.

**Comment by mlyszczek** [comment]
@onegray Thank you for your review. Good findings!

How familliar are you with both of stm32wl and stm32wb chips? If they are similar maybe we could think about merging these two chips into single stm32wx?

I see that stm32wl5 is mostly similar to stm32l4 chip.

**Comment by onegray** [comment]
> How familliar are you with both of stm32wl and stm32wb chips? If they are similar maybe we could think about merging these two chips into single stm32wx?
I have no experience with stm32wl and studying stm32wb during last year.  I think WB and WL should have separate implementations.
> I see that stm32wl5 is mostly similar to stm32l4 chip
stm32wb is very close to L4 as well, but there are still enough differences.

**Review by onegray — APPROVED** [review]
LGTM

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**